build-logic/jvm/src/main/kotlin/gradlebuild/jvm/extension/UnitTestAndCompileExtension.kt

```diff
@@ -51,6 +53,9 @@ abstract class UnitTestAndCompileExtension(
     private
     fun enforceCompatibility(majorVersion: Int) {
         tasks.withType<JavaCompile>().configureEach {
+            javaCompiler.set(project.javaToolchains.compilerFor {
+                languageVersion.set(JavaLanguageVersion.of(majorVersion))
+            })
```

Comment:
Previously `(source|target)Compatilbity` was enough to compile e.g., Java 6 code.

This is not true anymore; for example, JDK17 minimum supported version is JDK8, which means we have to use toolchains to select an appropriate JDK, which can build whatever version we require.

Reply:
🤔 This might change the build so that it requires 3 or 4 toolchains to run `compileAll` and not reuse compiler daemons as much. We could consider only using toolchains if `majorVersion < 8`.